WordPress/gutenberg
# Changelog tool: Enhance entry normalizations

This pull request seeks to enhance the plugin changelog generator tool to further enhance normalizations applied to release note entries:

- Omit entries by label (e.g. "Mobile App Compatibility", "Project Management")
- Omit by title prefix (e.g. "[RNMobile]")
- Rename headings (e.g. "Bug" → "Bug Fixes")
- Group related headings (e.g. "Bug" and "Regression" → "Bug Fixes")
- Order headings (e.g. "Features" at the top)
- Reword common technical terms (e.g. "e2e" → "end-to-end")
- Capitalize first meaningful segment (e.g. "e2e-tests: fix bug" → "e2e-tests: Fix bug")
   - Note: I considered to try to remove these prefixing, but I opted against it, since they can be useful for manual curation of entries, to reorganize entries to appropriate headings. This could be further enhanced in the future.

**Edit:** More enhancements added 2020-05-26:

- Derive type from pull request title (e.g. "Fixes ..." -> "Bug Fixes")
- Prioritizes assigned groups (e.g. "Enhancement" over "Task"/"Various")
- Remove redundant prefixes from title (e.g. remove prefix from "Code Quality: ..." if group is already known as "Code Quality")

All of the above were implemented in a way to facilitate future enhancements, e.g. to make it easy to omit more entries by label name.

**Testing Instructions:**

Ensure unit tests pass:

```
npm run test-unit bin/plugin/commands/test/changelog.js
```

Generate a changelog to assess the output is sensible:

```
npm run changelog
```
